### PR TITLE
bpo-43224: Implement substitution of unpacked TypeVarTuple

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue43224

### CODE REVIEW

File: Lib/typing.py [hunk 1]
```diff
@@ -1281,10 +1281,8 @@ def _determine_new_args(self, args):
         # anything more exotic than a plain `TypeVar`, we need to consider
         # edge cases.
 
-        if any(isinstance(p, TypeVarTuple) for p in self.__parameters__):
-            raise NotImplementedError(
-                "Type substitution for TypeVarTuples is not yet implemented"
-            )
+        if len(self.__parameters__) == 1 and isinstance(self.__parameters__[0], TypeVarTuple):
```
Comment (mrahtz) [inline Lib/typing.py]: I'm not sure it's sufficient to only check for the case where `len(self.__parameters__) == 1` - we could have e.g.:

```python
T1 = TypeVar('T1')
T2 = TypeVar('T2')
Ts = TypeVarTuple('Ts')
class A(Generic[T1, T2, Unpack[Ts]]): pass
B = A[int, T2, Unpack[Ts]]
C = B[str, float]
print(C)
```

There, `B.__parameters__` is `(T2, Ts)`, causing a `TypeError` later on in the code.

**Comment by mrahtz** [comment]
Thanks for looking into this!

File: Lib/typing.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/typing.py]: You are right. The previous code only worked with existing tests. New code adds tests for generics with multiple type variables.

**Comment by mrahtz** [comment]
Woah, I'm not sure merging this was a good idea. We discussed this extensively in https://github.com/python/cpython/pull/31021#pullrequestreview-890627941 and decided against implementing it this way. There are too many edge cases (e.g. the merged implementation will happily assign an unpacked arbitrary-length tuple such as `*tuple[int, ...]` to a plain `TypeVar`, which is not legal - see the extensive tests in https://github.com/python/cpython/commit/b9b1c80913f1fbc21ea60257bed6562e18c49802#diff-04d29c98076c2d6bb75921ea9becb26a862544d39b71db87b6e354c759b9305dL794), and @JelleZijlstra suspected it might cause issues with `Annotated`.

I'd strongly prefer this merge to be reversed in favour of https://github.com/python/cpython/pull/31804 and/or further discussion on what the right approach is.

**Comment by mrahtz** [comment]
Woah, I'm not sure merging this was a good idea. We discussed this extensively in https://github.com/python/cpython/pull/31021#pullrequestreview-890627941 and decided against implementing it this way. There are too many edge cases (e.g. the merged implementation will happily assign an unpacked arbitrary-length tuple such as `*tuple[int, ...]` to a plain `TypeVar`, which is not legal - see the extensive tests in https://github.com/python/cpython/commit/b9b1c80913f1fbc21ea60257bed6562e18c49802#diff-04d29c98076c2d6bb75921ea9becb26a862544d39b71db87b6e354c759b9305dL794), and @JelleZijlstra suspected it might cause issues with `Annotated`.

I'd strongly prefer this merge to be reversed in favour of https://github.com/python/cpython/pull/31804 and/or further discussion on what the right approach is.

**Comment by serhiy-storchaka** [comment]
I think that `list[T][int]` should return `list[int]`, and `tuple[*Ts][int, str]` should return `tuple[int, str]`.

Workaround can be used in cases where this does not work.

**Comment by mrahtz** [comment]
If you're adamant about keeping this implementation, I don't feel happy about it, but I don't want this to become a blocker for us. I'll submit a new PR with extra test cases so we can be more confident we've covered the edge cases.

**Pull request merged**